ValueTracking: simplify udiv/urem recurrences

artagnon opened the pull request:
A urem recurrence has the property that the result can never exceed the start value. A udiv recurrence has the property that the result can never exceed either the start value or the numerator, whichever is greater. Implement a simplification based on these properties.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1514,9 +1514,26 @@ static void computeKnownBitsFromOperator(const Operator *I,
+        if (Known2.isNonNegative())
+          Known.Zero.setHighBits(Known2.countMinLeadingZeros());
+        else if (Known2.isNegative())
+          Known.One.setHighBits(Known2.countMinLeadingOnes());
```
Is it correct? A counterexample: `for (int i = -9; cond; i /= 7)`.
The phi node should be known to be non-zero. I prefer to drop the `Known2.isNegative()` case.

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1451,34 +1451,34 @@ static void computeKnownBitsFromOperator(const Operator *I,
-        computeKnownBits(L, DemandedElts, Known3, Depth + 1, RecQ);
+      KnownBits Known3(BitWidth);
+      RecQ.CxtI = LInst;
+      computeKnownBits(L, DemandedElts, Known3, Depth + 1, RecQ);
```
the recursive calls should probably be guarded by opcode matches ot avoid unnecessary compile time overhead.

artagnon replied:
I think `matchSimpleRecurrence` already matches the Opcodes, and this code is guarded by `matchSimpleRecurrence`.

artagnon replied:
Actually, you're right; there's an additional FMul, for which we have no recurrence simplification.

artagnon commented:
Sorry, this is incorrect, and there are subtle problems with it. Is it possible to experiment with an Alive2 proof for this example? (I believe Alive2 doesn't handle loops)

dtcxzyw commented:
> Sorry, this is incorrect, and there are subtle problems with it. Is it possible to experiment with an Alive2 proof for this example? (I believe Alive2 doesn't handle loops)

You can try with `-src-unroll/-tgt-unroll`.

artagnon committed:
ValueTracking: simplify udiv/urem recurrences

udiv and urem recurrences have the property that the result can never
exceed the start value. Implement a simplification based on this
property.

artagnon commented:
As far as I can tell, there is no possible sdiv/srem recurrence simplification: with negative values, the result alternates between positive and negative values. I've now implemented just udiv/urem simplification, and I think it is simple enough that the patch should be good.

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -8997,12 +9012,14 @@ bool llvm::matchSimpleRecurrence(const PHINode *P, BinaryOperator *&BO,
     default:
       continue;
-    // TODO: Expand list -- xor, div, gep, uaddo, etc..
+    // TODO: Expand list -- sdiv, srem, fadd etc.
```
why remove the `xor` and `gep` and `uaddo` todos

artagnon replied:
- I can't think of a possible recurrence with xor.
- GEP isn't a BinaryOperator, is it? What possible recurrences are possible?
- I can't find uaddo in the LangRef, and don't know what it is.

goldsteinn replied:
> * I can't think of a possible recurrence with xor.
> 
Isn't that also true for `sdiv`/`srem`?

>     * GEP isn't a BinaryOperator, is it? What possible recurrences are possible?
> 
If it matches `ptradd` probably.

>     * I can't find uaddo in the LangRef, and don't know what it is.

Probably uadd_sat.

dtcxzyw commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1543,6 +1543,21 @@ static void computeKnownBitsFromOperator(const Operator *I,
         break;
       }
+
+      // Check for operations with the propperty that the magnitude of the
```
probability/property?

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1543,6 +1543,21 @@ static void computeKnownBitsFromOperator(const Operator *I,
+      // Check for operations with the property that the magnitude of the
+      // result will never exceed that of the start value.
+      case Instruction::UDiv:
+      case Instruction::URem: {
```
Isn't this missing a check for the operand?

(a udiv b) is <= a, but *not* <= b.

artagnon replied:
Doesn't `matchSimpleRecurrence` set `R` to the start value? I've also checked for `RInst` to set the context correctly.

artagnon commented:
Gentle ping.

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1543,6 +1543,29 @@ static void computeKnownBitsFromOperator(const Operator *I,
+          break;
+        [[fallthrough]];
+      }
+
```
No need to create a scope for `UDiv` case.

goldsteinn commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1543,6 +1543,29 @@ static void computeKnownBitsFromOperator(const Operator *I,
+
+        RecQ.CxtI = RInst;
+        computeKnownBits(R, DemandedElts, Known2, Depth + 1, RecQ);
+        Known.Zero.setHighBits(Known2.countMinLeadingZeros());
```
is `urem` where `phi` is on lhs also handled already?

goldsteinn replied:
Is there a reason you don't also try to get information from `lhs`. For `udiv` and `urem` can't exceed `lhs` either.

artagnon replied:
Isn't the BTC of the loop unknown (since we can't use SCEV), and hence is possibly-zero? In this case, will the PHI not take the initial value?

goldsteinn replied:
I don't follow, if its zero and we get wrong analysis is fine to handle the UB however we like. If its non-zero the analysis holds no?

artagnon replied:
Consider the following loop:

```llvm
loop:
  %iv = phi i64 [9, %entry], [%iv.next, %loop]
  %iv.next = urem i64 3, %iv
  %exitcond = icmp eq %iv.next, 3
  br i1 %exitcond, label %exit, label %loop
```

It has trip-count = 1, backedge-taken-count = 0. It is not UB. Do you agree that the maximum value of %iv is 9, and not 3?

goldsteinn replied:
How does that differ from RHS vs LHS?

artagnon replied:
I don't check RHS or LHS explicitly at all:

```cpp
        RecQ.CxtI = RInst;
        computeKnownBits(R, DemandedElts, Known2, Depth + 1, RecQ);
        Known.Zero.setHighBits(Known2.countMinLeadingZeros());
```

Here, `RInst` and `R` correspond to the PHI's start value.

goldsteinn replied:
Oh I see, I got myself a bit confused. This makes sense. Sorry for the confusion.

artagnon commented:
Gentle ping.

artagnon commented:
This patch is a very minor improvement, but an improvement nevertheless. I don't see it regressing compile-time, so it should be good, no?

nikic reviewed (commented):
Can this code be integrated in the shift handling above? It seems pretty similar -- but not quite the same, and I'm suspicious of differences :)

artagnon committed:
ValueTracking: integrate udiv/urem with shift code

artagnon commented:
> Can this code be integrated in the shift handling above? It seems pretty similar -- but not quite the same, and I'm suspicious of differences :)

The only difference is a smarter context. Given that the context is a little difficult to test, I've integrated the code at the cost of a worse context. We can try to improve the context in a follow-up, but I'm not sure what tests we can write.

nikic commented on llvm/lib/Analysis/ValueTracking.cpp:
```diff
@@ -1427,16 +1427,23 @@ static void computeKnownBitsFromOperator(const Operator *I,
+        [[fallthrough]];
 
+      // For a urem recurrence, the result can never exceed the start value. The
+      // start value could either be the numerator or the denominator.
```
The second sentence should probably refer to the phi / recurrence rather than start value?

nikic commented on llvm/test/Analysis/ValueTracking/recurrence-knownbits.ll:
```diff
@@ -81,6 +81,69 @@ exit:
+  ret i64 %res
+}
+
+define i64 @test_urem(i1 %c) {
```
Should test urem with the IV on LHS, as you also handle that case.

artagnon replied:
When IV is on LHS, it is already simplified (not by the recurrence code).

nikic reviewed (approved):
LGTM with some nits.

artagnon committed:
ValueTracking: fix nit in comment
